File: telemetry/aitf/sdk/rust/src/exporters/immutable_log.rs
//! Immutable, hash-chained audit-log exporter.
//!
//! Writes [`AIBaseEvent`] values to an append-only JSONL log where each entry
//! embeds a SHA-256 hash of the previous entry, forming an unbroken chain. Any
//! modification to a historical entry invalidates all subsequent hashes, making
//! tampering detectable. Ported from the Go `exporters/immutable_log.go`
//! (genesis hash, `seq | timestamp | prev_hash | event_json` payload, entry
//! `{seq, timestamp, prev_hash, hash, event}`, and chain verification).
//!
//! Satisfies audit requirements for EU AI Act Article 12 (record-keeping),
//! NIST AI RMF GOVERN-1.5 (audit trail), SOC 2 CC8.1 (integrity), and ISO/IEC
//! 42001 (AI management records).

use std::fs::OpenOptions;
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use serde_json::Value;

use crate::ocsf::AIBaseEvent;

/// The genesis hash that seeds the chain (64 zeros).
pub const GENESIS_HASH: &str = "0000000000000000000000000000000000000000000000000000000000000000";

/// Computes the SHA-256 hash for a log entry.
fn compute_entry_hash(seq: u64, timestamp: &str, prev_hash: &str, event_json: &str) -> String {
    let payload = format!("{seq}|{timestamp}|{prev_hash}|{event_json}");
    let digest = Sha256::digest(payload.as_bytes());
    hex(&digest)
}

fn hex(bytes: &[u8]) -> String {
    let mut s = String::with_capacity(bytes.len() * 2);
    for b in bytes {
        s.push_str(&format!("{b:02x}"));
    }
    s
}

/// A single entry in the hash-chained log.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImmutableLogEntry {
    pub seq: u64,
    pub timestamp: String,
    pub prev_hash: String,
    pub hash: String,
    pub event: Value,
}

/// Result of an integrity check over an immutable log file.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct VerificationResult {
    pub valid: bool,
    pub entries_checked: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub first_invalid_seq: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expected_hash: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub found_hash: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub final_hash: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

/// Writes hash-chained, tamper-evident JSONL log entries for AI events.
#[derive(Debug, Clone)]
pub struct ImmutableLogExporter {
    log_file: PathBuf,
    prev_hash: String,
    seq: u64,
    event_count: u64,
}

impl ImmutableLogExporter {
    /// Creates an exporter for `log_file`, resuming the chain from an existing
    /// file when present.
    pub fn new(log_file: impl AsRef<Path>) -> std::io::Result<Self> {
        let path = log_file.as_ref().to_path_buf();
        if let Some(dir) = path.parent() {
            if !dir.as_os_str().is_empty() {
                std::fs::create_dir_all(dir)?;
            }
        }
        let mut exporter = ImmutableLogExporter {
            log_file: path,
            prev_hash: GENESIS_HASH.to_string(),
            seq: 0,
            event_count: 0,
        };
        exporter.resume_chain();
        Ok(exporter)
    }

    fn resume_chain(&mut self) {
        let file = match std::fs::File::open(&self.log_file) {
            Ok(f) => f,
            Err(_) => return,
        };
        let mut last_line = String::new();
        for line in BufReader::new(file).lines().map_while(Result::ok) {
            let trimmed = line.trim();
            if !trimmed.is_empty() {
                last_line = trimmed.to_string();
            }
        }
        if !last_line.is_empty() {
            if let Ok(entry) = serde_json::from_str::<ImmutableLogEntry>(&last_line) {
                self.prev_hash = entry.hash;
                self.seq = entry.seq + 1;
            }
        }
    }

    /// Appends a batch of events to the log, extending the hash chain.
    pub fn export(&mut self, events: &[AIBaseEvent]) -> std::io::Result<()> {
        let mut lines: Vec<String> = Vec::new();
        for event in events {
            let event_value: Value = serde_json::to_value(event)
                .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
            let event_json = serde_json::to_string(&event_value)
                .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;

            let timestamp = now_rfc3339_nanos();
            let entry_hash = compute_entry_hash(self.seq, &timestamp, &self.prev_hash, &event_json);
            let entry = ImmutableLogEntry {
                seq: self.seq,
                timestamp,
                prev_hash: self.prev_hash.clone(),
                hash: entry_hash.clone(),
                event: event_value,
            };
            let entry_line = serde_json::to_string(&entry)
                .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
            lines.push(entry_line);
            self.prev_hash = entry_hash;
            self.seq += 1;
            self.event_count += 1;
        }

        if lines.is_empty() {
            return Ok(());
        }

        let mut file = OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.log_file)?;
        for line in lines {
            file.write_all(line.as_bytes())?;
            file.write_all(b"\n")?;
        }
        file.sync_all()?;
        Ok(())
    }

    /// Total number of events written by this exporter.
    pub fn event_count(&self) -> u64 {
        self.event_count
    }

    /// The next sequence number that will be assigned.
    pub fn current_seq(&self) -> u64 {
        self.seq
    }

    /// The hash of the most recently written entry (or the genesis hash).
    pub fn current_hash(&self) -> &str {
        &self.prev_hash
    }
}

/// Verifies the integrity of an immutable log file by replaying the hash chain
/// from genesis. Returns a [`VerificationResult`] describing any break.
pub fn verify_immutable_log(log_file: impl AsRef<Path>) -> VerificationResult {
    let file = match std::fs::File::open(log_file.as_ref()) {
        Ok(f) => f,
        Err(e) => {
            return VerificationResult {
                valid: false,
                error: Some(format!("cannot open file: {e}")),
                ..Default::default()
            }
        }
    };

    let mut prev_hash = GENESIS_HASH.to_string();
    let mut entries_checked: u64 = 0;

    for line in BufReader::new(file).lines().map_while(Result::ok) {
        let trimmed = line.trim();
        if trimmed.is_empty() {
            continue;
        }
        let entry: ImmutableLogEntry = match serde_json::from_str(trimmed) {
            Ok(e) => e,
            Err(e) => {
                return VerificationResult {
                    valid: false,
                    entries_checked,
                    first_invalid_seq: Some(entries_checked),
                    error: Some(format!("invalid JSON at seq {entries_checked}: {e}")),
                    ..Default::default()
                }
            }
        };

        if entry.prev_hash != prev_hash {
            return VerificationResult {
                valid: false,
                entries_checked,
                first_invalid_seq: Some(entry.seq),
                expected_hash: Some(prev_hash),
                found_hash: Some(entry.prev_hash),
                error: Some(format!("chain break at seq {}: prev_hash mismatch", entry.seq)),
                ..Default::default()
            };
        }

        let event_json = serde_json::to_string(&entry.event).unwrap_or_default();
        let computed = compute_entry_hash(entry.seq, &entry.timestamp, &entry.prev_hash, &event_json);
        if computed != entry.hash {
            return VerificationResult {
                valid: false,
                entries_checked,
                first_invalid_seq: Some(entry.seq),
                expected_hash: Some(computed),
                found_hash: Some(entry.hash),
                error: Some(format!("hash mismatch at seq {}: entry tampered", entry.seq)),
                ..Default::default()
            };
        }

        prev_hash = entry.hash;
        entries_checked += 1;
    }

    VerificationResult {
        valid: true,
        entries_checked,
        final_hash: Some(prev_hash),
        ..Default::default()
    }
}

fn now_rfc3339_nanos() -> String {
    use std::time::{SystemTime, UNIX_EPOCH};
    let d = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default();
    // Dependency-free RFC3339-style marker with nanosecond precision (mirrors
    // the crate's existing timestamp convention).
    format!("1970-01-01T00:00:00Z+{}.{:09}", d.as_secs(), d.subsec_nanos())
}

File: telemetry/aitf/sdk/python/src/aitf/exporters/immutable_log.py
"""AITF Immutable Log Exporter.

OTel SpanExporter that writes AI telemetry events to an append-only,
hash-chained log file providing cryptographic tamper evidence.

Each log entry includes a SHA-256 hash of the previous entry, creating
an unbroken chain. Any modification to a historical entry invalidates
all subsequent hashes, making tampering detectable.

This satisfies audit requirements for:
    - EU AI Act Article 12 (record-keeping obligations)
    - NIST AI RMF GOVERN-1.5 (audit trail for AI decisions)
    - SOC 2 CC8.1 (change management and integrity)
    - ISO/IEC 42001 (AI management system records)

Log format (JSONL, one entry per line):
    {
        "seq": 1,
        "timestamp": "2026-02-16T12:00:00.000Z",
        "prev_hash": "0000000000000000000000000000000000000000000000000000000000000000",
        "hash": "a1b2c3...",
        "event": { ... OCSF event ... }
    }

Verification:
    The integrity of the entire log can be verified by replaying the
    hash chain from entry 0 (genesis). Use ``ImmutableLogVerifier.verify()``
    to check an existing log file.

Usage:
    from aitf.exporters.immutable_log import ImmutableLogExporter

    exporter = ImmutableLogExporter(
        log_file="/var/log/aitf/immutable_audit.jsonl",
        compliance_frameworks=["eu_ai_act", "nist_ai_rmf", "soc2"],
    )
    provider = TracerProvider()
    provider.add_span_processor(BatchSpanProcessor(exporter))
"""

from __future__ import annotations

import hashlib
import json
import logging
import os
import threading
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Sequence

from opentelemetry.sdk.trace import ReadableSpan
from opentelemetry.sdk.trace.export import SpanExporter, SpanExportResult

from aitf.ocsf.compliance_mapper import ComplianceMapper
from aitf.ocsf.mapper import OCSFMapper

logger = logging.getLogger(__name__)

# Genesis hash (the "previous hash" for the very first entry)
_GENESIS_HASH = "0" * 64

# Maximum file size before rotation (1 GB)
_MAX_FILE_SIZE_BYTES = 1024 * 1024 * 1024

# Maximum file size to read during chain resume (100 MB)
# Prevents DoS if a very large file is placed at the log path
_MAX_RESUME_FILE_SIZE = 100 * 1024 * 1024


def _compute_entry_hash(
    seq: int,
    timestamp: str,
    prev_hash: str,
    event_json: str,
) -> str:
    """Compute SHA-256 hash for a log entry.

    The hash covers the sequence number, timestamp, previous hash,
    and the canonical JSON of the event. This ensures any change to
    any field in any entry breaks the chain.
    """
    payload = f"{seq}|{timestamp}|{prev_hash}|{event_json}"
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


class ImmutableLogExporter(SpanExporter):
    """Writes AITF events to an append-only, hash-chained log file.

    Each entry includes a SHA-256 hash linking it to the previous entry.
    The log file is opened in append-only mode with exclusive locking
    to prevent concurrent writes from corrupting the chain.

    Args:
        log_file: Path to the immutable log file.
        compliance_frameworks: Compliance frameworks for event enrichment.
        rotate_on_size: Enable automatic rotation when file exceeds 1 GB.
        file_permissions: UNIX file permissions (default: owner read/write).
    """

    def __init__(
        self,
        log_file: str = "/var/log/aitf/immutable_audit.jsonl",
        compliance_frameworks: list[str] | None = None,
        rotate_on_size: bool = True,
        file_permissions: int = 0o600,
    ) -> None:
        self._log_file = self._validate_log_path(log_file)
        self._rotate_on_size = rotate_on_size
        self._file_permissions = file_permissions
        self._mapper = OCSFMapper()
        self._compliance_mapper = ComplianceMapper(
            frameworks=compliance_frameworks
        )
        self._prev_hash = _GENESIS_HASH
        self._seq = 0
        self._lock = threading.Lock()
        self._event_count = 0

        # Create directory if needed
        self._log_file.parent.mkdir(parents=True, exist_ok=True)

        # Resume chain from existing log file
        self._resume_chain()

    @staticmethod
    def _validate_log_path(log_file: str) -> Path:
        """Validate the log file path to prevent path traversal.

        Rejects any path containing '..' components, which could be
        used to escape the intended directory.
        """
        raw_path = Path(log_file)

        if ".." in raw_path.parts:
            raise ValueError(
                f"Path traversal detected in log path: {log_file!r} "
                f"contains '..' component"
            )

        return raw_path.resolve()

    def _resume_chain(self) -> None:
        """Resume hash chain from existing log file if present."""
        if not self._log_file.exists():
            return

        # Guard against resuming from an excessively large file (DoS)
        file_size = self._log_file.stat().st_size
        if file_size > _MAX_RESUME_FILE_SIZE:
            logger.warning(
                "Log file %s is too large to resume (%d bytes > %d). "
                "Starting fresh chain.",
                self._log_file, file_size, _MAX_RESUME_FILE_SIZE,
            )
            return

        try:
            last_line = ""
            with open(self._log_file, "r") as f:
                for line in f:
                    line = line.strip()
                    if line:
                        last_line = line
                        self._seq += 1

            if last_line:
                entry = json.loads(last_line)
                self._prev_hash = entry.get("hash", _GENESIS_HASH)
                self._seq = entry.get("seq", self._seq) + 1
                logger.info(
                    "Resumed immutable log chain at seq=%d from %s",
                    self._seq, self._log_file,
                )
        except (json.JSONDecodeError, KeyError) as exc:
            logger.warning(
                "Could not resume chain from %s: %s. Starting fresh.",
                self._log_file, exc,
            )
            self._prev_hash = _GENESIS_HASH
            self._seq = 0

    def export(self, spans: Sequence[ReadableSpan]) -> SpanExportResult:
        """Export spans as hash-chained log entries."""
        entries: list[str] = []

        for span in spans:
            ocsf_event = self._mapper.map_span(span)
            if ocsf_event is None:
                continue

            # Enrich with compliance metadata
            event_type = self._classify_event(ocsf_event)
            if event_type:
                self._compliance_mapper.enrich_event(ocsf_event, event_type)

            event_dict = ocsf_event.model_dump(exclude_none=True)
            event_json = json.dumps(event_dict, sort_keys=True, default=str)

            with self._lock:
                timestamp = datetime.now(timezone.utc).isoformat()
                entry_hash = _compute_entry_hash(
                    self._seq, timestamp, self._prev_hash, event_json
                )

                entry = {
                    "seq": self._seq,
                    "timestamp": timestamp,
                    "prev_hash": self._prev_hash,
                    "hash": entry_hash,
                    "event": event_dict,
                }

                entries.append(
                    json.dumps(entry, sort_keys=False, default=str)
                )
                self._prev_hash = entry_hash
                self._seq += 1
                self._event_count += 1

        if not entries:
            return SpanExportResult.SUCCESS

        try:
            self._write_entries(entries)
            return SpanExportResult.SUCCESS
        except Exception as exc:
            logger.error("Immutable log write failed: %s", exc)
            return SpanExportResult.FAILURE

    def _write_entries(self, entries: list[str]) -> None:
        """Write entries to log file with rotation check."""
        # Check for rotation
        if (
            self._rotate_on_size
            and self._log_file.exists()
            and self._log_file.stat().st_size > _MAX_FILE_SIZE_BYTES
        ):
            rotated = self._log_file.with_suffix(
                f".{datetime.now(timezone.utc).strftime('%Y%m%dT%H%M%S')}.jsonl"
            )
            self._log_file.rename(rotated)
            logger.info("Rotated immutable log to %s", rotated)

        # Open in append-only mode with restrictive permissions
        fd = os.open(
            str(self._log_file),
            os.O_WRONLY | os.O_CREAT | os.O_APPEND,
            self._file_permissions,
        )
        try:
            with os.fdopen(fd, "a") as f:
                for entry in entries:
                    f.write(entry + "\n")
                f.flush()
                os.fsync(f.fileno())
        except Exception:
            # fd is already closed by os.fdopen context manager on error
            raise

    def _classify_event(self, event: Any) -> str | None:
        class_uid = getattr(event, "class_uid", 0)
        mapping = {
            7001: "model_inference",
            7002: "agent_activity",
            7003: "tool_execution",
            7004: "data_retrieval",
            7005: "security_finding",
            7006: "supply_chain",
            7007: "governance",
            7008: "identity",
        }
        return mapping.get(class_uid)

    @property
    def event_count(self) -> int:
        with self._lock:
            return self._event_count

    @property
    def current_seq(self) -> int:
        with self._lock:
            return self._seq

    @property
    def current_hash(self) -> str:
        with self._lock:
            return self._prev_hash

    def shutdown(self) -> None:
        pass

    def force_flush(self, timeout_millis: int = 30000) -> bool:
        return True


class ImmutableLogVerifier:
    """Verifies the integrity of an AITF immutable log file.

    Replays the hash chain from the genesis entry and validates that
    every entry's hash matches its computed value and links correctly
    to the previous entry.

    Usage:
        verifier = ImmutableLogVerifier("/var/log/aitf/immutable_audit.jsonl")
        result = verifier.verify()
        if result.valid:
            print(f"Log integrity verified: {result.entries_checked} entries")
        else:
            print(f"TAMPER DETECTED at seq {result.first_invalid_seq}")
            print(f"  Expected: {result.expected_hash}")
            print(f"  Found:    {result.found_hash}")
    """

    def __init__(self, log_file: str) -> None:
        self._log_file = Path(log_file)

    def verify(self) -> VerificationResult:
        """Verify the entire hash chain.

        Returns:
            VerificationResult with validation details.
        """
        if not self._log_file.exists():
            return VerificationResult(
                valid=False,
                entries_checked=0,
                error="Log file does not exist",
            )

        prev_hash = _GENESIS_HASH
        entries_checked = 0

        try:
            with open(self._log_file, "r") as f:
                for line_num, line in enumerate(f, 1):
                    line = line.strip()
                    if not line:
                        continue

                    try:
                        entry = json.loads(line)
                    except json.JSONDecodeError as exc:
                        return VerificationResult(
                            valid=False,
                            entries_checked=entries_checked,
                            first_invalid_seq=entries_checked,
                            error=f"Line {line_num}: invalid JSON: {exc}",
                        )

                    seq = entry.get("seq", -1)
                    timestamp = entry.get("timestamp", "")
                    stored_prev_hash = entry.get("prev_hash", "")
                    stored_hash = entry.get("hash", "")
                    event = entry.get("event", {})

                    # Check chain linkage
                    if stored_prev_hash != prev_hash:
                        return VerificationResult(
                            valid=False,
                            entries_checked=entries_checked,
                            first_invalid_seq=seq,
                            expected_hash=prev_hash,
                            found_hash=stored_prev_hash,
                            error=(
                                f"Chain break at seq {seq}: "
                                f"prev_hash mismatch"
                            ),
                        )

                    # Recompute hash
                    event_json = json.dumps(
                        event, sort_keys=True, default=str
                    )
                    computed_hash = _compute_entry_hash(
                        seq, timestamp, stored_prev_hash, event_json
                    )

                    if computed_hash != stored_hash:
                        return VerificationResult(
                            valid=False,
                            entries_checked=entries_checked,
                            first_invalid_seq=seq,
                            expected_hash=computed_hash,
                            found_hash=stored_hash,
                            error=(
                                f"Hash mismatch at seq {seq}: "
                                f"entry has been tampered with"
                            ),
                        )

                    prev_hash = stored_hash
                    entries_checked += 1

        except Exception as exc:
            return VerificationResult(
                valid=False,
                entries_checked=entries_checked,
                error=f"Verification failed: {exc}",
            )

        return VerificationResult(
            valid=True,
            entries_checked=entries_checked,
            final_hash=prev_hash,
        )


class VerificationResult:
    """Result of an immutable log verification."""

    def __init__(
        self,
        valid: bool,
        entries_checked: int = 0,
        first_invalid_seq: int | None = None,
        expected_hash: str | None = None,
        found_hash: str | None = None,
        final_hash: str | None = None,
        error: str | None = None,
    ) -> None:
        self.valid = valid
        self.entries_checked = entries_checked
        self.first_invalid_seq = first_invalid_seq
        self.expected_hash = expected_hash
        self.found_hash = found_hash
        self.final_hash = final_hash
        self.error = error

    def __repr__(self) -> str:
        if self.valid:
            return (
                f"VerificationResult(valid=True, "
                f"entries={self.entries_checked}, "
                f"final_hash={self.final_hash[:16]}...)"
            )
        return (
            f"VerificationResult(valid=False, "
            f"entries={self.entries_checked}, "
            f"error={self.error!r})"
        )
